core/src/broadcast_stage/broadcast_duplicates_run.rs

```diff
@@ -47,70 +55,19 @@ impl BroadcastDuplicatesRun {
             recent_blockhash: None,
             prev_entry_hash: None,
             num_slots_broadcasted: 0,
+            cluster_nodes_cache,
         }
     }
 
-    fn get_non_partitioned_batches(
-        &self,
-        my_pubkey: &Pubkey,
-        bank: &Bank,
-        data_shreds: Arc<Vec<Shred>>,
-    ) -> TransmitShreds {
-        let bank_epoch = bank.get_leader_schedule_epoch(bank.slot());
-        let mut stakes: HashMap<Pubkey, u64> = bank.epoch_staked_nodes(bank_epoch).unwrap();
-        stakes.retain(|pubkey, _stake| pubkey != my_pubkey);
-        (Some(Arc::new(stakes)), data_shreds)
-    }
-
     fn get_partitioned_batches(
         &self,
-        my_pubkey: &Pubkey,
         bank: &Bank,
         original_shreds: Arc<Vec<Shred>>,
         partition_shreds: Arc<Vec<Shred>>,
     ) -> (TransmitShreds, TransmitShreds) {
-        // On the last shred, partition network with duplicate and real shreds
-        let bank_epoch = bank.get_leader_schedule_epoch(bank.slot());
-        let mut original_recipients = HashMap::new();
-        let mut partition_recipients = HashMap::new();
-
-        let mut stakes: Vec<(Pubkey, u64)> = bank
-            .epoch_staked_nodes(bank_epoch)
-            .unwrap()
-            .into_iter()
-            .filter(|(pubkey, _)| pubkey != my_pubkey)
-            .collect();
-        stakes.sort_by(|(l_key, l_stake), (r_key, r_stake)| {
-            if r_stake == l_stake {
-                l_key.cmp(r_key)
-            } else {
-                l_stake.cmp(r_stake)
-            }
-        });
-
-        let mut cumulative_stake: u64 = 0;
-        for (pubkey, stake) in stakes.into_iter() {
-            cumulative_stake += stake;
-            if cumulative_stake <= self.config.stake_partition {
-                partition_recipients.insert(pubkey, stake);
-            } else {
-                original_recipients.insert(pubkey, stake);
-            }
-        }
-
-        warn!(
-            "{} sent duplicate slot {} to nodes: {:?}",
-            my_pubkey,
-            bank.slot(),
-            &partition_recipients,
-        );
-
-        let original_recipients = Arc::new(original_recipients);
-        let original_transmit_shreds = (Some(original_recipients), original_shreds);
-
-        let partition_recipients = Arc::new(partition_recipients);
-        let partition_transmit_shreds = (Some(partition_recipients), partition_shreds);
-
+        // TODO: This is no longer doing what it is used to do!
```

Comment:
Selfishly, it would be nice to maintain this old behavior if possible, since I have a few pending changes based on this 😃.

If you already have an idea of how to fixup this behavior, feel free to share. If not, I might have to take some creative freedoms to get this old behavior back 😈

Reply:
Awesome, thanks!

Reply:
@behzadnouri  I'm currently working to fix this up for `BroadcastDuplicatesRun`, will let you know if I run into any issues.

Reply:
Hey, this is the code to create partition on the last data shreds: https://github.com/behzadnouri/solana/commit/broadcast-duplicate

Feel free to use that code, or I can just send it as a pull-request to you. Either way is fine with me.

Please note that the old code:
https://github.com/solana-labs/solana/blob/b67ffab37/core/src/broadcast_stage/broadcast_duplicates_run.rs#L77-L99
was aiming to create a partition by manipulating the stakes effectively setting some of them to zero. But (even before all my changes to turbine) this does not actually create a partition. Because unstaked nodes were always part of the turbine broadcast and retransmit stages as long as their contact-info exists in gossip (ie. in cluster-info).

To actually create a partition you need to intercept where packets are sent over the socket as it is done here: https://github.com/behzadnouri/solana/commit/broadcast-duplicate

I am actively testing changes on turbine so if you plan to make changes here, please let me know in advance so that we do not step on each others' foot. cheers.

Reply:
sent out https://github.com/solana-labs/solana/pull/19253

Reply:
@behzadnouri thanks!